### ISSUE TITLE
Setting podMetadata labels without creation timestamp fails from golang

### ISSUE THREAD

**Issue by sverch** [coreos/prometheus-operator#2399]
**What did you do?**

Attempted to use the `podMetadata` field to attach labels to a Prometheus cluster from golang.

**What did you expect to see?**

The cluster get created with those labels attached.

**What did you see instead? Under which circumstances?**

The cluster fails to create with an error of `spec.podMetadata.creationTimestamp in body must be of type string: \"null\""`.

Full log line, showing the payload:

```
time="2019-02-12T23:00:26Z" level=error msg="Failed to deploy Prometheus" error="Prometheus.monitoring.coreos.com \"example-uswest\" is invalid: []: Invalid value: map[string]interface {}{\"apiVersion\":\"monitoring.coreos.com/v1\", \"kind\":\"Prometheus\", \"metadata\":map[string]interface {}{\"creationTimestamp\":\"2019-02-12T23:00:26Z\", \"labels\":map[string]interface {}{\"app\":\"example\"}, \"ownerReferences\":[]interface {}{map[string]interface {}{\"name\":\"example\", \"uid\":\"748b7e3a-2f19-11e9-b986-803820c5878b\", \"controller\":true, \"blockOwnerDeletion\":true, \"apiVersion\":\"example.com/v1alpha1\", \"kind\":\"ExampleCluster\"}}, \"generation\":1, \"uid\":\"fc290f8c-2f19-11e9-b986-803820c5878b\", \"name\":\"example-uswest\", \"namespace\":\"default\"}, \"spec\":map[string]interface {}{\"serviceAccountName\":\"prometheus\", \"serviceMonitorSelector\":map[string]interface {}{\"matchLabels\":map[string]interface {}{\"app\":\"example\"}}, \"podMetadata\":map[string]interface {}{\"creationTimestamp\":interface {}(nil), \"labels\":map[string]interface {}{\"monitoredApp\":\"example\"}}, \"resources\":map[string]interface {}{\"requests\":map[string]interface {}{\"memory\":\"1Gi\"}}}}: validation failure list:\nspec.podMetadata.creationTimestamp in body must be of type string: \"null\""
```

**Environment**

* Prometheus Operator version:

        `quay.io/coreos/prometheus-operator:v0.17.0`

* Kubernetes version information:

```
Client Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.1", GitCommit:"eec55b9ba98609a46fee712359c7b5b365bdd920", GitTreeState:"clean", BuildDate:"2018-12-13T10:39:04Z", GoVersion:"go1.11.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.0", GitCommit:"ddf47ac13c1a9483ea035a79cd7c10005ff21a6d", GitTreeState:"clean", BuildDate:"2018-12-03T20:56:12Z", GoVersion:"go1.11.2", Compiler:"gc", Platform:"linux/amd64"}
```

* Kubernetes cluster kind: 

Local cluster created using minikube.

* Manifests:

Code that fails:

```        
return prom.Prometheus{
                TypeMeta: metav1.TypeMeta{
                        Kind:       "Prometheus",
                        APIVersion: "monitoring.coreos.com/v1",
                },
                ObjectMeta: metav1.ObjectMeta{
                        Name:      "prometheus",
                        Labels: map[string]string{
                                "app": "example",
                        },
                },
                Spec: prom.PrometheusSpec{
                        Resources: v1.ResourceRequirements{
                                Requests: v1.ResourceList{
                                        v1.ResourceMemory: *resource.NewQuantity(1024*1024*1024, resource
.BinarySI),
                                },
                        },
                        ServiceMonitorSelector: &metav1.LabelSelector{
                                MatchLabels: map[string]string{
                                        "app": "example",
                                },
                        },
                        // validation failure list:\nspec.podMetadata.creationTimestamp in body must be of type string: \"null\""
                        PodMetadata: &metav1.ObjectMeta{
                                Labels: map[string]string{
                                        // We can't use "app", because the Prometheus operator
                                        // overrides that label with "prometheus" even if we set it
                                        // here.
                                        "monitoredApp": "example",
                                },
                                //CreationTimestamp: metav1Time,
                        },
                        ServiceAccountName: "prometheus",
                },
}
```

When I pass in a time of this type: https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/time.go, everything works (and I think it might eventually get overridden anyway).  I would expect this to be set by kubernetes.

* Prometheus Operator Logs:

Nothing much shows up there.  I think the request fails before it reaches the operator:

```
ts=2019-02-12T22:56:26.909458796Z caller=operator.go:258 component=prometheusoperator msg="connection established" cluster-version=v1.13.0
ts=2019-02-12T22:56:26.91212914Z caller=operator.go:158 component=alertmanageroperator msg="connection established" cluster-version=v1.13.0
ts=2019-02-12T22:56:26.913926233Z caller=operator.go:172 component=alertmanageroperator msg="CRD API endpoints ready"
ts=2019-02-12T22:56:26.916079722Z caller=operator.go:272 component=prometheusoperator msg="CRD API endpoints ready"
```

**Comment by zuzzas** [comment]
Suddenly started happening to our installations as well, albeit on a different fields. There is something wrong with CRD Validation. Why are nilable Kubernetes API types forced to be a string or an integer? Also, why are we even validating meta/v1 fields? cert-manager [is not doing this](https://github.com/jetstack/cert-manager/blob/master/deploy/manifests/00-crds.yaml).

```
metadata.deletionGracePeriodSeconds in body must be of type integer: "null"
metadata.deletionTimestamp in body must be of type string: "null"
```

@brancz, can you take a look?

**Comment by brancz** [comment]
You're using a version of the Prometheus Operator that is from 2017, could you please first upgrade and then verify? (both Kubernetes and Prometheus Operator have changed significantly since then)

**Comment by zuzzas** [comment]
@brancz 
I am on v0.28.0. Should I create a separate issue for my case?

**Comment by brancz** [comment]
Are both of you trying to programmatically create Prometheus objects? I think that might be the crooks of it, if you are.

**Comment by zuzzas** [comment]
Well, I am using Helm to apply a PrometheusRule CR. I am not specifying any of the "Kubernetes fields", such as deletionTimestamp, finalizers, etc.

I assume, they should appear in Kubernetes with their empty values values set. The empty value for [deletionGracePeriodSeconds](https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/types.go#L204) will be nil. Why does a validation fail? I am 100% sure that I'm missing something in the concept of CRD validation, can you help me find the way, please?

**Comment by brancz** [comment]
cc @s-urbaniak I believe this is related to the issue you are looking into regarding the volumeClaimTemplate.

**Comment by s-urbaniak** [comment]
@brancz yes definitely, this smells like the very same issue :+1: I am looking into it.

**Comment by zuzzas** [comment]
@s-urbaniak 
I can arrange an SSH access to a cluster that exhibits the problem, so that you can inspect API objects state in detail.
If you are interested, contact me via the e-mail on my profile.

**Comment by s-urbaniak** [comment]
@zuzzas Thanks a lot for the offer! I think I have a good understanding of the underlying issue now. The problem is that the generated CRD OpenAPI schema declares the creationTimestamp field as a simple string type:

```
                creationTimestamp:
                  format: date-time
                  type: string
```

But if this value is unset, it has a zero-value and is serialized as a `null` literal for JSON:
https://github.com/kubernetes/kubernetes/blob/9e4f8d6fae3a43833dbe9edcefd9170aa97496d7/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/time.go#L146-L149

This causes the CRD validation to fail as `null` is not a valid JSON string.

We have a couple of options:

1. We don't actually need the metadata fields from the embedded kubernetes resource types (i.e. for your failing `podMetadata`, but we have the same issue with other native resource types too). We could try to omit those somehow and generate CRDs referencing _only_ fields of interest.
2. Persuade the CRD generator to permit date-time strings to be nullable as per https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md#schemaNullable
3. Patch upstream to serialize `v1.Time` into empty string `""` in case of a zero value. This one obviously is hard to fix here in prometheus operator.

I will try out options 1. and 2.

@sverch In the meantime: Can you set creationTimestamp to `v1.Now()` as per https://github.com/kubernetes/kubernetes/blob/9e4f8d6fae3a43833dbe9edcefd9170aa97496d7/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/time.go#L61 for the time being and report if you have any issues with that?

**Comment by sverch** [comment]
Thanks @s-urbaniak!  Sorry for the slow reply.  I was able to set it to exactly that and it worked:

```
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
...
				CreationTimestamp: metav1.Now(),
```

**Comment by brancz** [comment]
We will need to follow up in the prometheus operator (CRDs) itself to fix this, as this should not be necessary.

**Comment by s-urbaniak** [comment]
It turns out, configuring the spec to enable a field to be nullable is actually not implemented currently in k8s. I paired with @sttts and he prepared the following PR https://github.com/kubernetes/kubernetes/pull/74804.

What this means in the short-term for the prometheus operator that the most viable option is to replicate the embedded core k8s types, stripping unnecassry fields from type/objectmeta and thus ommitting the need for nullability.

**Comment by brancz** [comment]
This is still a problem as far as I am aware.

**Comment by vsliouniaev** [comment]
Related to https://github.com/coreos/prometheus-operator/issues/2824

**Comment by brancz** [comment]
cc @pgier @s-urbaniak i believe this should now be resolved with the new controller generation tools no?

**Comment by s-urbaniak** [comment]
We should still verify that the new controller generation tools respect the `nullable` attribute. If they do, we're good here.

**Comment by Kampe** [comment]
This is still a problem

**Comment by s-urbaniak** [comment]
cc @pgier ^ this is something we should look into :thinking:

**Comment by pgier** [comment]
This may be fixed now by #3043 which replaces the ObjectMeta type for the pod metadata with a simpler type containing only labels and annotations.

**Comment by s-urbaniak** [comment]
Indeed @pgier that was a very neat workaround :+1: we will test this in our cluster-monitoring stack too. As there is no `creationTimestamp` at all any more, this should have resolved the issue.
